Mathlib/Data/UInt.lean

```diff
@@ -1,10 +1,231 @@
+
+      lemma add_def (a b : $typeName) : a + b = ⟨a.val + b.val⟩ := rfl
+
+      lemma eq_of_val_eq : ∀ {a b : $typeName}, a.val = b.val -> a = b
+      | ⟨f1⟩, ⟨f2⟩, h => congrArg mk h
+
+      lemma val_eq_of_eq : ∀ {a b : $typeName}, a = b -> a.val = b.val
+      | ⟨f1⟩, ⟨f2⟩, h => congrArg val h
+
+      @[simp] lemma mk_val_eq (a : $typeName) : mk a.val = a := by
+        cases h:a with
+        | mk aFin =>
+          have _ : a.val = aFin := h ▸ rfl
+          simp only [‹a.val = aFin›]
+
+      lemma add_assoc (a b c : $typeName) : (a + b) + c = a + (b + c) :=
+        congrArg mk (Fin.add_assoc _ _ _)
+
+      lemma mul_comm (a b : $typeName) : a * b = b * a := congrArg mk (Fin.mul_comm _ _)
+
+      lemma add_comm (a b : $typeName) : a + b = b + a := congrArg mk (Fin.add_comm _ _)
+
+      @[simp] lemma add_zero (a : $typeName) : a + 0 = a := by
+        have h0 : a.val + (0 : $typeName).val = a.val := AddMonoid.add_zero a.val
+        simp only [add_def, h0, mk_val_eq]
+
+      @[simp] lemma zero_add (a : $typeName) : 0 + a = a :=
+        (add_comm _ _) ▸ add_zero _
+
+      @[simp] lemma zero_mul (a : $typeName) : 0 * a = 0 := by
+        have h0 : (0 : $typeName).val * a.val = (0 : Fin size) := MonoidWithZero.zero_mul a.val
+        simp only [mul_def, h0, mk_val_eq]
+
+      @[simp] lemma mul_zero (a : $typeName) : a * 0 = 0 :=
+        (mul_comm a _) ▸ zero_mul _
+
+      @[simp] lemma one_mul (a : $typeName) : 1 * a = a := by
+        have h0 : (1 : $typeName).val * a.val = a.val := Monoid.one_mul a.val
+        simp only [mul_def, h0, mk_val_eq]
+
+      @[simp] lemma mul_one (a : $typeName) : a * 1 = a := (mul_comm _ _) ▸ one_mul _
+
+      instance : Neg $typeName where
+        neg a := mk (-a.val)
+
+      lemma neg_def (a : $typeName) : -a = mk (-a.val) := rfl
+
+      lemma sub_eq_add_neg (a b : $typeName) : a - b = a + -b :=
+        congrArg mk (SubNegMonoid.sub_eq_add_neg _ _)
+
+      def nsmul (x : Nat) (a : $typeName) : $typeName := mk (AddMonoid.nsmul x a.val)
+
+      def gsmul (x : Int) (a : $typeName) : $typeName := mk (SubNegMonoid.gsmul x a.val)
+
+      @[simp] lemma gsmul_zero' (a : $typeName) : gsmul 0 a = (0 : $typeName) :=
+        congrArg mk (SubNegMonoid.gsmul_zero' a.val)
+
+      lemma gsmul_succ' (m : ℕ) (a : $typeName) : gsmul (Int.ofNat m.succ) a = a + gsmul (Int.ofNat m) a :=
+        congrArg mk (SubNegMonoid.gsmul_succ' m a.val)
+
+      lemma gsmul_neg' (m : ℕ) (a : $typeName) : gsmul (Int.negSucc m) a = -(gsmul m.succ a) :=
+        congrArg mk (SubNegMonoid.gsmul_neg' m a.val)
+
+      instance : AddSemigroup $typeName where
+        add_assoc := add_assoc
+
+      @[simp] lemma nsmul_zero' (x : $typeName) : nsmul 0 x = 0 :=
+        congrArg mk (AddMonoid.nsmul_zero' x.val)
+
+      lemma nsmul_succ' (n : ℕ) (x : $typeName) : nsmul n.succ x = x + nsmul n x :=
+        congrArg mk (AddMonoid.nsmul_succ' n x.val)
+
+      @[simp] lemma add_left_neg (a : $typeName) : -a + a = 0 :=
```

Comment:
This is already a simp lemma from the Group type class, we shouldn't add duplicate simp lemmas for every type.
```suggestion
      lemma add_left_neg (a : $typeName) : -a + a = 0 :=
```

It's also never used, and therefore should be removed entirely I think.